clang/lib/CodeGen/CodeGenModule.cpp

```diff
@@ -3554,16 +3566,9 @@ void CodeGenModule::EmitGlobal(GlobalDecl GD) {
       // device-side variables because the CUDA runtime needs their
       // size and host-side address in order to provide access to
       // their device-side incarnations.
-
-      // So device-only functions are the only things we skip, except for SYCL.
-      if (isa<FunctionDecl>(Global) && !Global->hasAttr<CUDAHostAttr>() &&
+      if (!LangOpts.isSYCL() && isa<FunctionDecl>(Global) &&
```

Comment:
Please refactor this code. You don't need the `else` block anymore since all it seems to do is `return`. Also, shouldn't the assert on L3547 be before the `if `check?

Reply:
Please forgive me in case I'm misunderstanding your comment.

I think we need the `else` block in line 3564 for the purpose of skipping the global emission in case of the host compilation of CUDA (only) codes. This is basically what the upstream llvm does.
https://github.com/intel/llvm/blob/ec782951d7bd34f9c32d08bb6ec6b43d2a805832/clang/lib/CodeGen/CodeGenModule.cpp#L3249-L3263

I would like to leave the assert in line 3574 in that place for the same reason.